contrib/tools/meta.md/mtransform.md

```diff
@@ -89,57 +88,107 @@ Removing Conditional Equations/Rules
-    if S' := qid("C" + string(S)) .
-
-    op unconditionalize : Sort Sort Qid Qid Module -> [Module] .
-    ------------------------------------------------------------
-    eq unconditionalize(S, CS, OP, Q, M) = fromTemplate(getName(M), unconditionalize(S, CS, OP, Q, asTemplate(M))) .
-
     op stripConditions : ModuleDeclSet -> [ModuleDeclSet] .
     -------------------------------------------------------
-    eq stripConditions(IS SDS SSDS OPDS MAS EQS)    = IS SDS SSDS OPDS MAS EQS .
+    eq stripConditions(IDS SDS SSDS OPDS MAS EQS)   = IDS SDS SSDS OPDS MAS EQS .
     eq stripConditions(NeMDS NeMDS')                = stripConditions(NeMDS) stripConditions(NeMDS') .
     eq stripConditions(  rl T => T'      [ AS ] . ) = ( rl T => T' [ AS ] . ) .
     eq stripConditions( crl T => T' if C [ AS ] . ) = ( rl T => T' [ AS ] . ) .
+
+    op stripConditions : Module -> [Module] .
+    -----------------------------------------
+    eq stripConditions(M) = fromTemplate(getName(M), stripConditions(asTemplate(M))) .
+
+    op conditionFor : Qid ModuleDeclSet -> [Condition] .
+    ----------------------------------------------------
+    eq conditionFor(Q, crl T => T' if C [ label(Q) AS ] . MDS) = C   .
+    eq conditionFor(Q,  rl T => T'      [ label(Q) AS ] . MDS) = nil .
+
+    op conditionFor : Qid Module -> [Condition] .
+    ---------------------------------------------
+    eq conditionFor(Q, M) = conditionFor(Q, asTemplate(M)) .
+```
+
+### Unconditionalize
+
+The general algorithm for unconditionalizing rules with equational conditions follows.
+It's parametric in several operators and sorts (prefixed below with `#`) about the module which the conditions will be translated into.
+
+```maude
+    op unconditionalize : ModuleDeclSet -> [ModuleDeclSet] .
+    --------------------------------------------------------
+   ceq unconditionalize(IDS SDS SSDS OPDS MAS EQS RLS)
+     = IDS  ( pr #cModule . )
+       SDS  ( sorts CS . )
+       SSDS
+       OPDS ( op #cTerm : #tSort #cSort -> CS [ctor prec(57)] . )
+       MAS
+       EQS
+       unconditionalizeRules(RLS)
+    if CS := qid("C" + string(#tSort)) .
+
+    op unconditionalize : Module -> [Module] .
+    ------------------------------------------
+    eq unconditionalize(M) = fromTemplate(getName(M), unconditionalize(asTemplate(M))) .
+
+    op unconditionalizeRules : RuleSet -> [RuleSet] .
+    -------------------------------------------------
+    eq unconditionalizeRules(none)         = none .
+    eq unconditionalizeRules(NeRLS NeRLS') = unconditionalizeRules(NeRLS) unconditionalizeRules(NeRLS') .
+
+   ceq unconditionalizeRules(rl T => T' [AS] .) = ( rl #cTerm[T, V] => #cTerm[T', V] [narrowing AS] . )
+    if V := qid("###COND###:" + string(#cSort)) .
+
+   ceq unconditionalizeRules(crl T => T' if EqC [AS] .) = ( rl #cTerm[T, V] => #cTerm[T', C'] [narrowing AS] . )
+    if V  := qid("###COND###:" + string(#cSort))
+    /\ C' := #cConjunct[V, #mkCondition(EqC)] .
+```
```

Comment:
Should unconditionalize() be limited only to rules with the narrowing attribute?

Reply:
Well, we want it to apply to all rules I think. We can't have `narrowing` on conditional rules because Maude complains, and if we don't inject the `narrowing` attribute in the resulting theory, it won't be used for narrowing. So the idea was that we unconditionalize all rules so that they all can be used for narrowing uniformly.

Reply:
Sorry, I missed the fact that the `narrowing` is added later. I'm in full agreement here.